### PR TITLE
Add separated decoder_head_mask for T5 Models

### PR DESCRIPTION
**Pull request by stancld**
### Fix issue #9632 

<hr>

This PR separates `head_mask` and `decoder_head_mask` for T5 models, and thus enables to specify different head masks for an encoder and decoder.

**Description:**
- Replace a single input argument `head_mask` with a separated couple `head_mask` and `decoder_head_mask` for the T5 models: `T5Model, T5ForConditionalGeneration, TFT5Model, TFT5ForConditionalGeneration`
- Slightly change the order of input arguments to follow the convention of first 7 arguments introduced in PR #9569 for BART-based models, i.e. `"input_ids", "attention_mask", "decoder_input_ids", "decoder_attention_mask", "head_mask", "decoder_head_mask", "encoder_outputs"`
- Currently, the updated PyTorch T5 model does not pass `test_forward_signature` in `tests/test_modeling_common.py`. This problem will be diminished once PR #9569 to be merged.

Reviewer: @patrickvonplaten (the code is ready for review)

### CODE REVIEW

**Review by patrickvonplaten — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
Great, that looks nice! Let's first merge https://github.com/huggingface/transformers/pull/9569 and then rebase this PR so that it passes all tests :-)

**Review by LysandreJik — COMMENTED** [review]
Cool, welcome feature! 

This is a slight breaking change, isn't it? Users that once used the `head_mask` had it control both their encoder and their decoder. From now on, specifying `head_mask` only controls the encoder, leaving the decoder to be controlled by `decoder_head_mask`.

Can we do a deprecation cycle, where if no `decoder_head_mask` is given, we set it to the value of `head_mask`? Having a `FutureWarning` there would be nice, too.

**Comment by talkhaldi** [comment]
Thanks for fixing this!

I have one note/question: This seems to only apply to self-attention heads, not heads in the cross attention module, right? Is this intentional?

**Comment by stancld** [comment]
@talkhaldi Thank you very much for pointing this out. It seems you're right and this is not intentional by myself. It'll be fixed in another commit.

**Comment by stancld** [comment]
Hey @patrickvonplaten and @LysandreJik. I've added some `FutureWarning` into the code to handle cases when only `head_mask` is passed by a user. Also, I fixed a cross-attention issue noted by @talkhaldi.
I believe, the PR is now ready for review as all the checks have passed after the rebasing.

File: src/transformers/models/t5/modeling_t5.py [hunk 1]
```diff
@@ -1258,6 +1276,18 @@ def forward(
...
+        if head_mask is not None and decoder_head_mask is None:
+            if self.config.num_layers == self.config.num_decoder_layers:
+                warning_msg = """
+                The input argument `head_mask` was split into two arguments `head_mask` and `decoder_head_mask`.
+                Currently, `decoder_head_mask` is set to copy `head_mask`, but this feature is deprecated and will be
+                removed in future versions. If you do not want to use any `decoder_head_mask` now, please set
+                `decoder_head_mask = torch.ones(num_layers, num_heads)`.
+                """
+                warnings.warn(warning_msg, FutureWarning)
+                decoder_head_mask = head_mask
```
Comment (LysandreJik) [inline src/transformers/models/t5/modeling_t5.py:1289]: Great message!

**Review by LysandreJik — APPROVED** [review]
LGTM! Thanks for making the change @stancld.

File: src/transformers/models/t5/modeling_t5.py [hunk 2]
```diff
@@ -1447,6 +1479,18 @@ def forward(
         use_cache = use_cache if use_cache is not None else self.config.use_cache
         return_dict = return_dict if return_dict is not None else self.config.use_return_dict
 
+        # FutureWarning: head_mask was separated into two input args - head_mask, decoder_head_mask
+        if head_mask is not None and decoder_head_mask is None:
+            if self.config.num_layers == self.config.num_decoder_layers:
+                warning_msg = """
```
Comment (sgugger) [inline src/transformers/models/t5/modeling_t5.py:1485]: I see this message is used twice, maybe it could be refactored in a private constant?

File: src/transformers/models/t5/modeling_tf_t5.py [hunk 3]
```diff
@@ -1289,16 +1319,28 @@ def call(
             >>> result = model.generate(inputs)
 
         """
+        # FutureWarning: head_mask was separated into two input args - head_mask, decoder_head_mask
+        if head_mask is not None and decoder_head_mask is None:
+            warning_msg = """
```
Comment (sgugger) [inline src/transformers/models/t5/modeling_tf_t5.py:1324]: Same here.

**Review by sgugger — APPROVED** [review]
Thanks for your PR, it's very clean!

File: src/transformers/models/t5/modeling_t5.py (hunk 2 above)
Comment (stancld) [reply src/transformers/models/t5/modeling_t5.py]: That's a good point, thanks!
